akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/SlotState.scala

```diff
@@ -0,0 +1,195 @@
+ *
+ * Implementation of slot logic that is completed decoupled from the machinery bits which are implemented in the GraphStageLogic
+ * and exposed only through [[SlotContext]].
+ */
+@InternalApi
+private[pool] object SlotState {
+  sealed abstract class ConnectedState extends SlotState {
+    //def connection: SlotConnection
+    def isConnected: Boolean = true
+  }
+  sealed trait IdleState extends SlotState {
+    final override def isIdle = true
+  }
+  sealed trait BusyState extends SlotState {
+    final override def isIdle = false // no HTTP pipelining right now
+    def ongoingRequest: RequestContext
+  }
+
+  case object Unconnected extends SlotState with IdleState {
+    def isConnected: Boolean = false
+
+    override def onPreConnect(ctx: SlotContext): SlotState = {
+      ctx.openConnection()
+      PreConnecting
+    }
+
+    override def onNewRequest(ctx: SlotContext, requestContext: RequestContext): SlotState = {
+      ctx.openConnection()
+      Connecting(requestContext)
+    }
+  }
+  case object Idle extends ConnectedState with IdleState with WithRequestDispatching {
+    override def onNewRequest(ctx: SlotContext, requestContext: RequestContext): SlotState =
+      dispatchRequestToConnection(ctx, requestContext)
+
+    override def onConnectionCompleted(ctx: SlotContext): SlotState = Unconnected
+    override def onConnectionFailed(ctx: SlotContext, cause: Throwable): SlotState = Unconnected
+  }
+  sealed trait WithRequestDispatching { _: ConnectedState ⇒
+    def dispatchRequestToConnection(ctx: SlotContext, ongoingRequest: RequestContext): SlotState = {
+      val r = ongoingRequest.request
+      ctx.pushRequestToConnectionAndThen(r, WaitingForEndOfRequestEntity(ongoingRequest))
+    }
+  }
+
+  final case class Connecting(ongoingRequest: RequestContext) extends ConnectedState with BusyState with WithRequestDispatching {
+    override def onConnectedAttemptSucceeded(ctx: SlotContext, outgoingConnection: Http.OutgoingConnection): SlotState = {
+      ctx.debug("Slot connection was established")
+      dispatchRequestToConnection(ctx, ongoingRequest)
+    }
+    override def onConnectionAttemptFailed(ctx: SlotContext, cause: Throwable): SlotState = {
+      ctx.debug("Connection attempt failed.")
+      // FIXME: register failed connection attempt, schedule request for rerun, backoff new connection attempts
+      ???
+    }
+  }
+
+  // FIXME: unify with Connecting
+  case object PreConnecting extends ConnectedState with IdleState with WithRequestDispatching {
+    override def onConnectedAttemptSucceeded(ctx: SlotContext, outgoingConnection: Http.OutgoingConnection): SlotState = {
+      ctx.debug("Slot connection was (pre-)established")
+      Idle
+    }
+    override def onConnectionAttemptFailed(ctx: SlotContext, cause: Throwable): SlotState = {
+      ctx.debug("Connection attempt failed.")
+      // FIXME: register failed connection attempt, schedule request for rerun, backoff new connection attempts
```

Comment:
so slot context needs a way to schedule a thing, and that'd call into openConnection I guess?
State should become something signifying that we're in this backoff period I guess?

Reply:
We might add a dedicated `SlotState` but the important thing here is that the backoff state should be shared between all slots.